coalib/core/Core.py

```diff
@@ -0,0 +1,264 @@
+        cleanup_bear(bear, result_callback, running_tasks, event_loop,
+                     executor)
+
+    if results is not None:
+        for result in results:
+            try:
+                # FIXME Long operations on the result-callback could block the
+                # FIXME   scheduler significantly. It should be possible to
+                # FIXME   schedule new Python Threads on the given event_loop
+                # FIXME   and process the callback there.
+                result_callback(result)
+            except Exception as ex:
+                # FIXME Try to display only the relevant traceback of the
+                # FIXME result handler if error occurred there, not the
+                # FIXME complete event-loop traceback.
+                logging.error(
+                    'An exception was thrown during result-handling.',
+                    exc_info=ex)
+
+
+def initialize_dependencies(bears):
+    """
+    Initializes and returns a ``DependencyTracker`` instance together with a
+    set of bears ready for scheduling.
+
+    This function acquires, processes and registers bear dependencies
+    accordingly using a consumer-based system, where each dependency bear has
+    only a single instance per section and file-dictionary.
+
+    The bears set returned accounts for bears that have dependencies and
+    excludes them accordingly. Dependency bears that have themselves no further
+    dependencies are included so the dependency chain can be processed
+    correctly.
+
+    :param bears:
+        The set of bears to run that serve as an entry-point.
+    :return:
+        A tuple with ``(dependency_tracker, bears_to_schedule)``.
+    """
+    # Pre-collect bears in a set as we use them more than once. Especially
+    # remove duplicate instances.
+    bears = set(bears)
+
+    dependency_tracker = DependencyTracker()
+
+    # For a consumer-based system, we have a situation which can be visualized
+    # with a graph:
+    #
+    # (section1, file_dict1) (section1, file_dict2) (section2, file_dict2)
+    #       |       |                  |                      |
+    #       V       V                  V                      V
+    #     bear1   bear2              bear3                  bear4
+    #       |       |                  |                      |
+    #       V       V                  |                      |
+    #  BearType1  BearType2            -----------------------|
+    #       |       |                                         |
+    #       |       |                                         V
+    #       ---------------------------------------------> BearType3
+    #
+    # We need to traverse this graph and instantiate dependency bears
+    # accordingly, one per section.
+
+    # Group bears by sections and file-dictionaries. These will serve as
+    # entry-points for the dependency-instantiation-graph.
+    grouping = groupby(bears, key=lambda bear: (bear.section, bear.file_dict))
```

Comment:
Since we need to sort the data first, so it'd be a better efficient way to do the grouping with dict. like,
`def group_by(l, key=lambda..):
    d = defaultdict(list)
    for item in l:
        d[key(item)].append(item)
    return d.items()`